### ISSUE TITLE
Configuration Cache support

### ISSUE THREAD

**Issue by rahulsainani** [ben-manes/gradle-versions-plugin#666]
Got this warning in  Android Studio's build analyzer:

```
com.github.ben-manes.versions: not compatible 
You could save about 4.3s by turning Configuration cache  on. 
With Configuration cache, Gradle can skip the configuration phase entirely when nothing that affects the build configuration has changed.

The version of this plugin used in this build is not compatible with Configuration cache and we don’t know the version when it becomes compatible.  
 
 Plugin version: 0.42.0 
 Plugin dependency: com.github.ben-manes:gradle-versions-plugin
```

Is it possible to use configuration cache with this plugin?

**Comment by ben-manes** [comment]
Compatibility would require a rewrite to use per-project tasks and a root-level accumulator task (similar to jacoco). PRs are appreciated if someone wants to work on that. Instead we mark the task as not compatible so that the configuration cache ignores our task and lets you get the benefits in other builds.

**Comment by tprochazka** [comment]
Technically I think, that ignoring these plugin tasks until I call it is entirely enough. Because the most important is build speed during development and nobody actually needs to call this plugin as often as build&run operation. When you really want to check version updates, it is not so important that it will take a little bit longer.

Currently, last released version has this in the changelog
> Merge pull request https://github.com/ben-manes/gradle-versions-plugin/pull/585 from eygraber/dont-fail-w-configuration-cache
>       Prevent builds from failing w/ configuration cache enabled on Gradle 7.4+

So why does Android studio still display the warning that this plugin is blocking to use of compilation cache?

And there is also one possible workaround. It is possible detect that gradle is invoked from Android Studio IDE by this function

```
val Project.isIDEBuild: Boolean
    get() {
        return this.hasProperty("android.injected.invoked.from.ide")
    }

```
So you can apply the plugin only when the project is not invoked from IDE.

**Comment by jaredsburrows** [comment]
We would have to rewrite parts of this plugin to not pass `project` our `@TaskAction` to make `--configuration-cache` pass in a test.

**Comment by ianbrandt** [comment]
Testing upgrade to [Gradle 8.1 RC1](https://docs.gradle.org/8.1-rc-1/release-notes.html):
* If I run `dependencyUpdates` with `org.gradle.configuration-cache=true`, it fails with, "Invocation of 'Task.project' by task ':dependencyUpdates' at execution time is unsupported."
* If I add `org.gradle.configuration-cache.problems=warn`, the build doesn't fail, but `dependencyUpdates` reports, "No dependencies found."

One oddity here, without `org.gradle.configuration-cache.problems=warn`, I also get configuration cache errors for the `compileJava` and `jar` tasks in addition to the `dependencyUpdates` task when running `dependencyUpdates`. I don't get these additional errors when just running `build` or `jar`, or if I do add `org.gradle.configuration-cache.problems=warn`.

Reproducer with `org.gradle.configuration-cache.problems=warn` currently set: https://github.com/sdkotlin/sd-kotlin-talks/tree/gradle-8.1 (specifically, https://github.com/sdkotlin/sd-kotlin-talks/tree/73c97f04b1bbbe361ea6c51ae1886f70e208bfa0).

**Comment by ianbrandt** [comment]
Testing with [Gradle 8.1 RC2](https://docs.gradle.org/8.1-rc-2/release-notes.html):
* If I run `dependencyUpdates` with just `org.gradle.configuration-cache=true`, it no longer fails now that "gradle/gradle#24411" is fixed.
  * I do get, "47 problems were found storing the configuration cache, 5 of which seem unique.", as expected given this open issue, but they don't include any errors for the `compileJava` or `jar` tasks anymore.
* If I add `org.gradle.configuration-cache.problems=warn`, `dependencyUpdates` still reports, "No dependencies found."

It would be nice to be able to set `org.gradle.configuration-cache=true`, `org.gradle.configuration-cache.problems=warn`, and still be able to use this plugin.

Reproducer: https://github.com/sdkotlin/sd-kotlin-talks/tree/929b0ccfd607bb41019cce34accd4abf94f35408.

Should I file a separate issue for it, since it's not the same thing as full support for the configuration cache?

**Comment by ianbrandt** [comment]
I'd expect the plugin to still print available updates, even if `org.gradle.configuration-cache=true` and `org.gradle.configuration-cache.problems=warn` are set. With Gradle 8.1 RC2 it just prints, "No dependencies found."

**Comment by ben-manes** [comment]
Then that sounds like a Gradle bug if it broke between 8.0 and 8.1?

I upgraded and ran it on Caffeine where the report came out as expected. I'll try yours when I get some downtime, but I think you should report your findings for the Gradle team to hopefully fix during the RC process.

**Comment by ben-manes** [comment]
Running your reproducer and I think you should open a Gradle issue. I see that passing `--no-configuration-cache` will generate a report. This means all of our logic works correctly, except when caching is enabled in which case it must get empty results?

I am still trying to reproduce it outside of the plugin to understand where it deviates. I will likely have to build a custom version with logging to see where it breaks down, since it is very obscure.

Can you start a conversation with the Gradle folks?

**Comment by ben-manes** [comment]
Hacking up your build, and I believe the configuration cache is buggy when crossing the includeBuild boundary.

If I add a dependency in `platform:test-platform` and the `allConfigurationDependencies` task defined above then,
- `gradle platform:test-platform:allConfigurationDependencies` -> no results
- `cd platforms/test-platform && gradle allConfigurationDependencies` -> dependencies

hope that helps.

**Comment by ianbrandt** [comment]
Thanks for looking into the "No dependencies" issue! Initial comment posted to https://github.com/gradle/gradle/issues/24411#issuecomment-1491040354.

**Comment by ianbrandt** [comment]
Filed https://github.com/gradle/gradle/issues/24636.

**Comment by FunkyMuse** [comment]
**Stacktrace**
```
3 problems were found storing the configuration cache, 2 of which seem unique.
- Task `:dependencyUpdates` of type `com.github.benmanes.gradle.versions.updates.DependencyUpdatesTask`: cannot serialize a lambda that captures or accepts a parameter of type 'org.gradle.api.artifacts.Configuration' as these are not supported with the configuration cache.
  See https://docs.gradle.org/8.2/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
- Task `:dependencyUpdates` of type `com.github.benmanes.gradle.versions.updates.DependencyUpdatesTask`: invocation of 'Task.project' at execution time is unsupported.
  See https://docs.gradle.org/8.2/userguide/configuration_cache.html#config_cache:requirements:use_project_during_execution
```

```invocation of Task.project📋 at execution time is unsupported```

This is the error I get

**Comment by Vampire** [comment]
Why? Unless you degrade CC problems to warnings which should always only be a temporary measure during build-logic development, it should work fine, as the task declares that it is not CC-compatible which makes the CC-state being discarded before running the execution phase. Only if you degrade problems to warnings this is no longer the case and the CC-state is used even if there were problems or tasks declaring that they don't support CC, and this often causes problems then.

**Comment by yusu-netapp** [comment]
I'm trying to upgrade a project from gradle 7.4.2 to 8.10, and the configuration-cache is a must-have due to some internal requirements. I'm applying the dependency update task through a plugin so that it is applied to each subproject automatically. When I'm building, I get some error with this stacktrace:
```
Caused by: org.gradle.api.internal.plugins.PluginApplicationException: Failed to apply plugin 'org.gradle.help-tasks'.
        at org.gradle.api.internal.plugins.DefaultPluginManager.doApply(DefaultPluginManager.java:176)
        at org.gradle.api.internal.plugins.DefaultPluginManager.apply(DefaultPluginManager.java:148)
        at org.gradle.api.plugins.internal.HelpTasksAutoApplyAction.execute(HelpTasksAutoApplyAction.java:25)
        at org.gradle.api.plugins.internal.HelpTasksAutoApplyAction.execute(HelpTasksAutoApplyAction.java:22)
        at org.gradle.configuration.project.PluginsProjectConfigureActions.execute(PluginsProjectConfigureActions.java:47)
        at org.gradle.configuration.project.PluginsProjectConfigureActions.execute(PluginsProjectConfigureActions.java:25)
        at org.gradle.configuration.project.ConfigureActionsProjectEvaluator.evaluate(ConfigureActionsProjectEvaluator.java:35)
        at org.gradle.configuration.project.LifecycleProjectEvaluator$EvaluateProject.lambda$run$0(LifecycleProjectEvaluator.java:109)
        at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.lambda$applyToMutableState$1(DefaultProjectStateRegistry.java:411)
        at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.lambda$fromMutableState$2(DefaultProjectStateRegistry.java:434)
        at org.gradle.internal.work.DefaultWorkerLeaseService.withReplacedLocks(DefaultWorkerLeaseService.java:359)
        at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.fromMutableState(DefaultProjectStateRegistry.java:434)
        at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.applyToMutableState(DefaultProjectStateRegistry.java:410)
        at org.gradle.configuration.project.LifecycleProjectEvaluator$EvaluateProject.run(LifecycleProjectEvaluator.java:100)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:166)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:47)
        at org.gradle.configuration.project.LifecycleProjectEvaluator.evaluate(LifecycleProjectEvaluator.java:72)
        at org.gradle.api.internal.project.DefaultProject.evaluate(DefaultProject.java:768)
        at org.gradle.api.internal.project.DefaultProject.evaluate(DefaultProject.java:158)
        at org.gradle.api.internal.project.ProjectLifecycleController.lambda$ensureSelfConfigured$2(ProjectLifecycleController.java:85)
        at org.gradle.internal.model.StateTransitionController.lambda$doTransition$14(StateTransitionController.java:255)
        at org.gradle.internal.model.StateTransitionController.doTransition(StateTransitionController.java:266)
        at org.gradle.internal.model.StateTransitionController.doTransition(StateTransitionController.java:254)
        at org.gradle.internal.model.StateTransitionController.lambda$maybeTransitionIfNotCurrentlyTransitioning$10(StateTransitionController.java:199)
        at org.gradle.internal.work.DefaultSynchronizer.withLock(DefaultSynchronizer.java:34)
        at org.gradle.internal.model.StateTransitionController.maybeTransitionIfNotCurrentlyTransitioning(StateTransitionController.java:195)
        at org.gradle.api.internal.project.ProjectLifecycleController.ensureSelfConfigured(ProjectLifecycleController.java:85)
        at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.ensureConfigured(DefaultProjectStateRegistry.java:385)
        at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.ensureConfigured(DefaultProjectStateRegistry.java:383)
        at org.gradle.api.internal.project.DefaultProject.evaluationDependsOn(DefaultProject.java:846)
        at org.gradle.api.internal.project.DefaultProject.evaluationDependsOnChildren(DefaultProject.java:828)
        at com.github.benmanes.gradle.versions.updates.DependencyUpdatesTask.dependencyUpdates(DependencyUpdatesTask.kt:127)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:125)
```
Which becomes very confusing. I'm not familiar with kotlin, so I will try my best to read the code, but in the mean time can anyone give me some insights about what is this error about? Is it more because of the configuration cache or more because of the way I apply this plugin?

**Comment by ben-manes** [comment]
This plugin will aggregate child projects so you can apply it only at the root. That might fix your problem.

The failed line is a call to [project.evaluationDependsOnChildren()](https://docs.gradle.org/current/javadoc/org/gradle/api/Project.html#lifecycle-heading). This was added by a contributor in #83 for configuration-on-demand support, though I can't say I understood their reasoning. It might be that your build has a required evaluation order and that call forces an incompatible one. I think it would be fine to remove that line, but also if you only apply at the root it may solve your problems.

**Comment by Vampire** [comment]
I don't think that stacktrace has anything to do particularly with CC support, especially as the task declares it is not CC-compatible yet so CC is not used either way.

But I think that this `project.evaluationDependsOnChildren` call at task execution time is highly questionable either way.

Usually you call this in a build script to make the child projects be evaluated first before continuing with configuring the current project if you have some bad-practice cross-project configuration and similar. But doing this at execution time I even wonder that it actually has the intended effect, if it does have it actually.

But latest with proper CC support in this task, the question would be mood as you cannot use `project` at execution time anyway with CC on, and even with CC off it is deprecated to do so, so sooner or later the call has to go.

**Comment by yusu-netapp** [comment]
Thank you for the update!
I'd like to do it only once in the root, but this project is like a monorepo, and different team may work on a small subset of the subprojects only, so they prefer to have the dependencyUpdate reports in individual subprojects instead of the aggregated one in the root. 
I can definitely try only apply it in the root and see if I'm still hitting the build problem, but would prefer to keep current behavior if possible.
